types/wlr_xdg_shell.c

```diff
@@ -0,0 +1,1457 @@
+		free(surface->toplevel_state);
+	}
+
+	if (surface->role == WLR_XDG_SURFACE_ROLE_POPUP) {
+		wl_resource_set_user_data(surface->popup_state->resource, NULL);
+
+		if (surface->popup_state->seat) {
+			struct wlr_xdg_popup_grab *grab =
+				xdg_shell_popup_grab_from_seat(surface->client->shell,
+					surface->popup_state->seat);
+
+			struct wlr_xdg_surface *topmost =
+				xdg_popup_grab_get_topmost(grab);
+
+			if (topmost != surface) {
+				wl_resource_post_error(surface->client->resource,
+						XDG_WM_BASE_ERROR_NOT_THE_TOPMOST_POPUP,
+						"xdg_popup was destroyed while it was not the topmost "
+						"popup.");
+			}
+
+			wl_list_remove(&surface->popup_state->grab_link);
+
+			if (wl_list_empty(&grab->popups)) {
+				if (grab->seat->pointer_state.grab == &grab->pointer_grab) {
+					wlr_seat_pointer_end_grab(grab->seat);
+				}
+				if (grab->seat->keyboard_state.grab == &grab->keyboard_grab) {
+					wlr_seat_keyboard_end_grab(grab->seat);
+				}
+			}
+		}
+
+		wl_list_remove(&surface->popup_state->link);
+		free(surface->popup_state);
+	}
+
+	wl_resource_set_user_data(surface->resource, NULL);
+	wl_list_remove(&surface->link);
+	wl_list_remove(&surface->surface_destroy_listener.link);
+	wlr_surface_set_role_committed(surface->surface, NULL, NULL);
+	free(surface->geometry);
+	free(surface->next_geometry);
+	free(surface->title);
+	free(surface->app_id);
+	free(surface);
+}
+
+
+static void xdg_positioner_destroy(struct wl_resource *resource) {
+	struct wlr_xdg_positioner *positioner =
+		wl_resource_get_user_data(resource);
+	free(positioner);
+
+}
+
+static void xdg_positioner_protocol_set_size(struct wl_client *client,
+		struct wl_resource *resource, int32_t width, int32_t height) {
+	struct wlr_xdg_positioner *positioner =
+		wl_resource_get_user_data(resource);
+
+	if (width < 1 || height < 1) {
+		wl_resource_post_error(resource,
+			XDG_POSITIONER_ERROR_INVALID_INPUT,
+			"width and height must be positives and non-zero");
+		return;
+	}
+
+	positioner->size.width = width;
+	positioner->size.height = height;
+}
+
+static void xdg_positioner_protocol_set_anchor_rect(struct wl_client *client,
+		struct wl_resource *resource, int32_t x, int32_t y, int32_t width,
+		int32_t height) {
+	struct wlr_xdg_positioner *positioner =
+		wl_resource_get_user_data(resource);
+
+	if (width < 1 || height < 1) {
+		wl_resource_post_error(resource,
+			XDG_POSITIONER_ERROR_INVALID_INPUT,
+			"width and height must be positives and non-zero");
+		return;
+	}
+
+	positioner->anchor_rect.x = x;
+	positioner->anchor_rect.y = y;
+	positioner->anchor_rect.width = width;
+	positioner->anchor_rect.height = height;
+}
+
+static void xdg_positioner_protocol_set_anchor(struct wl_client *client,
+		struct wl_resource *resource, uint32_t anchor) {
+	struct wlr_xdg_positioner *positioner =
+		wl_resource_get_user_data(resource);
+
+	/* post error if anchor > XDG_POSITIONER_ANCHOR_BOTTOM_RIGHT ? */
```

Comment:
Should we check invalid value here?
Given it's an enum without a last enum value to use as marker, what's the style-wise correct way of checking bound?

Reply:
Yes, checking would be cool (please backport to xdg-shell unstable v6 as well).

We generally define a mask of valid bits (see https://github.com/swaywm/wlroots/blob/master/types/wlr_data_device.c#L13).

Reply:
Ah, if it isn't a bitmask anymore, it makes more sense to check if `anchor > XDG_POSITIONER_ANCHOR_BOTTOM_RIGHT`.

Reply:
xdg-shell unstable-v6 had the check, but the logic changed so I couldn't keep it as is.

I don't 100% like the `anchor > XDG_POSITIONER_ANCHOR_BOTTOM_RIGHT` construct for enum because if an extra enum value is added at the end it is hard to locate these in the code (that's why my enums tend to have a final _COUNT or similar at the end), but I guess that's a non-concern for a stable xdg-shell interface.
I'll add the check as is.